app/src/main/java/com/example/playlistmaker/SearchActivity.kt

```diff
@@ -25,33 +50,131 @@ class SearchActivity : AppCompatActivity() {
             finish()
         }
 
-        inputSearchField = findViewById<EditText>(R.id.searchEditText)
+        inputSearchField = findViewById(R.id.searchEditText)
 
         val clearImageView = findViewById<ImageView>(R.id.searchClearIcon)
         clearImageView.setOnClickListener {
-            inputSearchField?.setText(DEFAULT_TEXT)
+            inputSearchField.setText(DEFAULT_TEXT)
+            clearTrackList()
             val inputMethodManager =
                 getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
-            inputMethodManager?.hideSoftInputFromWindow(inputSearchField?.windowToken, 0)
+            inputMethodManager?.hideSoftInputFromWindow(inputSearchField.windowToken, 0)
         }
 
         val textWatcher = object : TextWatcherJustAfterTextChanged {
 
             override fun afterTextChanged(s: Editable?) {
                 clearImageView.isVisible = checkImageViewVisibility(s)
-                inputSearchText = inputSearchField?.text.toString()
+                inputSearchText = inputSearchField.text.toString()
             }
         }
 
-        inputSearchField?.addTextChangedListener(textWatcher)
-
-        val listOfTrackMock = ListOfTrackMock()
+        inputSearchField.addTextChangedListener(textWatcher)
 
         val tracksRecyclerView = findViewById<RecyclerView>(R.id.tracksRecyclerView)
         tracksRecyclerView.layoutManager =
             LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
-        val trackAdapter = TracksAdapter(listOfTrackMock.getList())
         tracksRecyclerView.adapter = trackAdapter
+
+        placeholderImage = findViewById(R.id.placeholderSearch)
+        errorTextTextView = findViewById(R.id.errorTextTextView)
+        refreshButton = findViewById(R.id.refreshButton)
+
+        inputSearchField.setOnEditorActionListener { _, actionId, _ ->
+            if (actionId == EditorInfo.IME_ACTION_DONE) {
+                search()
+                true
+            }
+            false
+        }
+
+        refreshButton.setOnClickListener {
+            search() }
+    }
+
+    private fun search() {
+        itunesService.search(inputSearchText).enqueue(object : Callback<TrackItunesResponse> {
+            override fun onResponse(
+                call: Call<TrackItunesResponse>,
+                response: Response<TrackItunesResponse>
+            ) {
+                when (response.code()) {
+                    200 -> {
+                        if (response.body()?.results?.isNotEmpty() == true) {
+                            val diffResult = getDiffResult(
+                                oldList = tracks,
+                                newList = response.body()!!.results
+                            )
+                            tracks.clear()
+                            tracks.addAll(response.body()?.results!!)
```

Comment:
!! - это тоже самое что кричать, лучше не кричать на компилятор, если response.body() будет равен null, то произойдёт краш, так как каждый раз вызывая response.body() ты можешь получить разные экезмпляры, то проверка выше тебе не поможет
```suggestion
   200 -> {
                        val body: TrackItunesResponse? = response.body()

                        if (body?.results?.isNotEmpty() == true) {
                            val diffResult = getDiffResult(
                                oldList = tracks,
                                newList = body.results
                            )

                            tracks.clear()
                            tracks.addAll(body.results)
                            diffResult.dispatchUpdatesTo(trackAdapter)
                            showMessage(ResponseStatusCodes.OK)
                        } else {
                            showMessage(ResponseStatusCodes.NOTHING_FOUND)
                        }
                    }
                    else -> showMessage(ResponseStatusCodes.NO_CONNECTION)
                }
```

Reply:
понял, спасибо за пояснение!